libraries/AP_DDS/AP_DDS_Client.cpp

```diff
@@ -276,35 +298,70 @@ void AP_DDS_Client::main_loop(void)
     }
 }
 
-
-bool AP_DDS_Client::init()
+bool AP_DDS_Client::ddsSerialInit()
 {
-    AP_SerialManager *serial_manager = AP_SerialManager::get_singleton();
-    dds_port = serial_manager->find_serial(AP_SerialManager::SerialProtocol_DDS_XRCE, 0);
-    if (dds_port == nullptr) {
-        return false;
-    }
-
-    // ensure we own the UART
-    dds_port->begin(0);
-
     constexpr uint8_t fd = 0;
     constexpr uint8_t relativeSerialAgentAddr = 0;
     constexpr uint8_t relativeSerialClientAddr = 1;
-    if (!uxr_init_serial_transport(&serial_transport,fd,relativeSerialAgentAddr,relativeSerialClientAddr)) {
+    if (!uxr_init_serial_transport(&serial_transport, fd, relativeSerialAgentAddr, relativeSerialClientAddr)) {
         return false;
     }
-
     constexpr uint32_t uniqueClientKey = 0xAAAABBBB;
     //TODO does this need to be inside the loop to handle reconnect?
     uxr_init_session(&session, &serial_transport.comm, uniqueClientKey);
+    return true;
+}
+
+#if AP_DDS_UDP_ENABLED
+bool AP_DDS_Client::ddsUdpInit()
+{
+    char port_string[10];
+    hal.util->snprintf(port_string, sizeof(port_string), "%u", udp_port.get());
+
+    if (!uxr_init_udp_transport(&udp_transport, UXR_IPv4, ip, port_string)) {
+        return false;
+    }
+    constexpr uint32_t uniqueClientKey = 0xAAAABBBB;
+    uxr_init_session(&session, &udp_transport.comm, uniqueClientKey);
+    return true;
+}
+#endif // AP_DDS_UDP_ENABLED
+
+bool AP_DDS_Client::init()
+{
+    // serial init will fail if the SERIALn_PROTOCOL is not setup
+    bool initTransportStatus = ddsSerialInit();
+    is_serial = initTransportStatus;
+
+#if AP_DDS_UDP_ENABLED
+    // fallback to UDP if available
+    if (!initTransportStatus) {
+        initTransportStatus = ddsUdpInit();
+    }
+#endif
+
+    if (!initTransportStatus) {
+        GCS_SEND_TEXT(MAV_SEVERITY_INFO,"DDS Client: Transport Initialization failed");
+        return false;
+    }
     while (!uxr_create_session(&session)) {
         GCS_SEND_TEXT(MAV_SEVERITY_INFO,"DDS Client: Initialization waiting...");
         hal.scheduler->delay(1000);
     }
 
-    reliable_in = uxr_create_input_reliable_stream(&session,input_reliable_stream,BUFFER_SIZE_SERIAL,STREAM_HISTORY);
-    reliable_out = uxr_create_output_reliable_stream(&session,output_reliable_stream,BUFFER_SIZE_SERIAL,STREAM_HISTORY);
+    reliable_in = uxr_create_input_reliable_stream(&session, input_reliable_stream, BUFFER_SIZE_SERIAL, STREAM_HISTORY);
+    reliable_out = uxr_create_output_reliable_stream(&session, output_reliable_stream, BUFFER_SIZE_SERIAL, STREAM_HISTORY);
+
+    if (is_serial) {
```

Comment:
Is the plan to only support one or the other? Would there be a strong reason to have redundant communication methods between ArduPilot and Companion computer?

Reply:
we could do both in the future, but for now it is simpler to have one